### PR TITLE
Thread safety in network models

### PR DESCRIPTION
**Pull request by sarahscott**
@alloy, this code is functioning well for me, and AFNetworking is indeed releasing the request upon its completion. Is this what you had in mind? 

TODO:
- [x] Discuss
- [x] Update code
- [x] Tests

### CODE REVIEW

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m [hunk 1]
```diff
@@ -59,41 +59,29 @@ - (void)getWorksForYou:(void (^_Nonnull)(NSArray<ARWorksForYouNotificationItem *
...
 {
-    if (self.downloadLock) {
+    NSAssert([NSThread isMainThread], @"This should only be called by the main thread");
+
+    if (self.currentRequest) {
         return;
     }
 
-    _downloadLock = YES;
-    __weak typeof(self) wself = self;
```
Comment (orta) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: I think you'll still want to do the weak/strong self dance here

Comment (alloy) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: For the most memory performant code, doing the dance means the network model’s memory can be released before the request is finished. But we’re not introducing a retain-cycle by not doing the dance, so maybe we can live with that for slightly more readable code?

Comment (orta) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: :+1:

**Comment by alloy** [comment]
@sarahscott Yeah this is exactly what I meant and good work on testing it gets released :+1: What do you think? To me the code looks much cleaner.

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m [hunk 2]
```diff
@@ -59,41 +59,29 @@ - (void)getWorksForYou:(void (^_Nonnull)(NSArray<ARWorksForYouNotificationItem *
...
 
     } failure:^(NSError *error) {
-        __strong typeof (wself) sself = wself;
-        if (!sself) return;
-
-        sself.allDownloaded = NO;
-        sself.downloadLock = NO;
 
+        self.allDownloaded = NO;
         success(@[]);
```
Comment (alloy) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: Shouldn’t you call the `failure` block here?

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m [hunk 3]
```diff
@@ -59,41 +59,29 @@ - (void)getWorksForYou:(void (^_Nonnull)(NSArray<ARWorksForYouNotificationItem *
...
+        success(artworks);
 
     } failure:^(NSError *error) {
-        __strong typeof (wself) sself = wself;
-        if (!sself) return;
-
-        sself.allDownloaded = NO;
-        sself.downloadLock = NO;
 
+        self.allDownloaded = NO;
```
Comment (alloy) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: While explicit, I don’t think this is really needed, the default value is already `NO`.

**Comment by sarahscott** [comment]
Yeah, it's a lot cleaner. I don't have as much experience with memory-related nightmares so of course I'm happier without the dance, but willing to put it back if it makes sense to do so.

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m (hunk 2 above)
Comment (sarahscott) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: Absolutely!

**Comment by alloy** [comment]
> I don't have as much experience with memory-related nightmares so of course I'm happier without the dance, but willing to put it back if it makes sense to do so.

It’s a tough :cookie:. _Always_ doing the dance is of course generally the safest thing to do. It might mean an object gets released before you’d expect, but you’d probably notice that and can introduce a strong reference at that time.

OTOH, it’s probably a good idea to understand the dangers fully, so maybe we should do a short sit-down on that when I’m over? (Remind me :wink:)

**Comment by sarahscott** [comment]
That'd be awesome. I'll make a reminder for that, and re-implement the dance for now.

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Warning</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td><p>PR is classed as Work in Progress</p>
</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/KrauseFx/danger/">danger</a>
</p>

**Comment by sarahscott** [comment]
I'm going to pair with @mennenia for a bit and then write tests for this.

**Comment by sarahscott** [comment]
Assigning to @orta because @alloy will be traveling this weekend - let me know if I should change that!

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m [hunk 4]
```diff
@@ -54,26 +60,22 @@ - (void)getWorksForYou:(void (^_Nonnull)(NSArray<ARWorksForYouNotificationItem *
         NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"date" ascending:NO];
         success([notificationItems sortedArrayUsingDescriptors:@[descriptor]]);
 
-    } failure:nil];
+    } failure:^(NSError *error) {
```
Comment (orta) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: you can pass the `failure` here, instead `failure: failure]` the blocks are the same type :+1:

File: Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m [hunk 5]
```diff
@@ -81,19 +83,9 @@ - (void)getArtworks:(void (^_Nonnull)(NSArray<Artwork *> *_Nonnull))success fail
...
     } failure:^(NSError *error) {
-        __strong typeof (wself) sself = wself;
-        if (!sself) return;
-
-        sself.allDownloaded = NO;
-        sself.downloadLock = NO;
-
-        success(@[]);
+        failure(error);
     }];
```
Comment (orta) [inline Artsy/Networking/Network_Models/ARWorksForYouNetworkModel.m]: ^ same

**Comment by orta** [comment]
Cool, looks good - you're good to :ship: - unless you want to make my failure fixes :+1:

**Comment by orta** [comment]
ace

**Pull request merged**